uaclient/files.py

```diff
@@ -0,0 +1,296 @@
+import json
+import logging
+import os
+from datetime import datetime
+from typing import Dict, Optional
+
+from uaclient import defaults, event_logger, schemas, util
+
+PRIVATE_SUBDIR = "private"
+
+event = event_logger.get_event_logger()
+LOG = logging.getLogger(__name__)
+
+
+class UAFile:
+    _root_directory = defaults.DEFAULT_DATA_DIR
+    _file_name = "file"
+
+    def __init__(
+        self,
+        root_directory: Optional[str] = None,
+        filename: Optional[str] = None,
+        private: bool = True,
+    ):
+        if root_directory:
+            self._root_directory = root_directory
+        if filename:
+            self._file_name = filename
+        self._is_private = private
+        self._filepath = None
+
+    @property
+    def filepath(self):
+        if not self._filepath:
+            self._filepath = os.path.join(
+                self._root_directory, self._file_name
+            )
+        return self._filepath
+
+    @property
+    def is_private_file(self):
+        return self._is_private
+
+    def write(self, content):
+        file_mode = 0o600 if self._is_private else 0o644
+        self.create_file()
+        if not isinstance(content, str):
+            content = json.dumps(content, cls=util.DatetimeAwareJSONEncoder)
+        util.write_file(self.filepath, content, file_mode)
```

Comment:
I'd rather keep this layer simple and only support string content
```suggestion
    def write(self, content: str):
        file_mode = 0o600 if self._is_private else 0o644
        self.create_file()
        util.write_file(self.filepath, content, file_mode)
```